source/simulator/assembly.cc

```diff
@@ -1701,6 +1807,839 @@ namespace aspect
+                       :
+                       0.0);
+                    const double latent_heat_LHS =
+                      ((advection_field.is_temperature())
+                       ?
+                       face_heating_model_outputs.lhs_latent_heat_terms[q]
+                       :
+                       0.0);
+                    Assert (density_c_P + latent_heat_LHS >= 0,
+                            ExcMessage ("The sum of density times c_P and the latent heat contribution "
+                                        "to the left hand side needs to be a non-negative quantity."));
+
+                    const double penalty = (advection_field.is_temperature()
+                                            ?
+                                            10
+                                            * parameters.temperature_degree
+                                            * parameters.temperature_degree
+                                            / face->measure()
+                                            * conductivity
+                                            / (density_c_P + latent_heat_LHS)
+                                            :
+                                            0.0);
+
+                    const double fixed_temperature_value = (advection_field.is_temperature()
+                                                            ?
+                                                            boundary_temperature->temperature(*geometry_model,
+                                                                                              cell->face(f)->boundary_id(),
+                                                                                              scratch.face_finite_element_values.quadrature_point(q))
+                                                            :
+                                                            0.0);
+
+                    Tensor<1,dim> current_u = scratch.face_current_velocity_values[q];
+                    //Subtract off the mesh velocity for ALE corrections if necessary
+                    if (parameters.free_surface_enabled)
+                      current_u -= scratch.face_mesh_velocity_values[q];
+
+                    for (unsigned int i=0; i<advection_dofs_per_cell; ++i)
+                      {
+                        data.local_rhs(i)
+                        += (- time_step *  conductivity
+                            * scratch.face_grad_phi_field[i]
+                            * scratch.face_finite_element_values.normal_vector(q)
+                            * fixed_temperature_value
+
+                            + time_step
+                            * (density_c_P + latent_heat_LHS)
+                            * penalty
+                            * scratch.face_phi_field[i]
+                            * fixed_temperature_value
+
+                            + (current_u * scratch.face_finite_element_values.normal_vector(q) < 0
```

Comment:
Also I think I remember we had a similar case with a different condition in the Stokes system, and evaluating this expression outside of the loops for matrix and rhs and using it as a constant bool inside actually saved quite some time in the assembly. I am not sure if the same applies here, but maybe it is worth the change, because then you also have more space to explain this expression.